libs/ml/tests/ml/model/sequential_test.cpp

```diff
@@ -0,0 +1,153 @@
+
+TYPED_TEST_CASE(ModelsTest, MyTypes);
+
+template <typename TypeParam>
+void PrepareTestDataAndLabels1D(TypeParam &train_data, TypeParam &train_label,
+                                TypeParam &test_datum, TypeParam &test_label)
+{
+  train_data  = TypeParam::FromString("1.1, 2.1, 3.1; 4.1, 5.1, 6.1; 7.1, 8.1, 9.1");
+  train_label = TypeParam::FromString("0.1; 1.1; 2.1");
+  test_datum  = TypeParam::FromString("10.1; 11.1; 12.1");
+  test_label  = TypeParam::FromString("3.1");
+}
+
+template <typename TypeParam, typename DataType, typename ModelType>
+ModelType SetupModel(fetch::ml::optimisers::OptimiserType     optimiser_type,
+                     fetch::ml::model::ModelConfig<DataType> &model_config, TypeParam &data,
+                     TypeParam &gt)
+{
+  // setup dataloader
+  using DataLoaderType = fetch::ml::dataloaders::TensorDataLoader<TypeParam, TypeParam>;
+  SizeVector              label_shape = {gt.shape().at(0), 1};
+  std::vector<SizeVector> data_shape  = {{data.shape().at(0), 1}};
+  auto data_loader_ptr                = std::make_unique<DataLoaderType>(label_shape, data_shape);
+  data_loader_ptr->AddData(data, gt);
+
+  // run model in training mode
+  auto model = ModelType(model_config);
+  model.template Add<fetch::ml::layers::FullyConnected<TypeParam>>(3, 100);
+  model.template Add<fetch::ml::layers::FullyConnected<TypeParam>>(100, 100);
+  model.template Add<fetch::ml::layers::FullyConnected<TypeParam>>(100, 1);
+  model.SetDataloader(std::move(data_loader_ptr));
+  model.Compile(optimiser_type);
+
+  return model;
+}
+
+template <typename TypeParam>
+bool RunTest(fetch::ml::optimisers::OptimiserType optimiser_type,
+             typename TypeParam::Type             tolerance,
+             typename TypeParam::Type             lr = static_cast<typename TypeParam::Type>(0.5))
+{
+  using DataType  = typename TypeParam::Type;
+  using ModelType = fetch::ml::model::Sequential<TypeParam>;
+
+  fetch::math::SizeType n_training_steps = 10;
+
+  fetch::ml::model::ModelConfig<DataType> model_config;
+  model_config.learning_rate_param.mode =
+      fetch::ml::optimisers::LearningRateParam<DataType>::LearningRateDecay::EXPONENTIAL;
+  model_config.learning_rate_param.starting_learning_rate = lr;
+  model_config.learning_rate_param.exponential_decay_rate = static_cast<DataType>(0.99);
+
+  // set up data
+  TypeParam train_data, train_labels, test_datum, test_label;
+  PrepareTestDataAndLabels1D<TypeParam>(train_data, train_labels, test_datum, test_label);
+
+  // set up model
+  ModelType model = SetupModel<TypeParam, DataType, ModelType>(optimiser_type, model_config,
+                                                               train_data, train_labels);
+
+  // test loss decreases
+  fetch::math::SizeType count{0};
+  while (count < n_training_steps)
+  {
+    DataType loss{0};
+    DataType later_loss{0};
+    EXPECT_TRUE(model.Train(1, loss));
+    EXPECT_TRUE(model.Train(1, later_loss));
```

Comment:
Should there be a check here that later_loss < loss?

Reply:
yep, nice catch